### PR TITLE
bmcdiscover -i IP -u USERID -p PASSW0RD --ipsource failed

### PR DESCRIPTION
**Pull request by cxhong**
for issue #1872,  the bmc_user and bmc_pass is uninitialized for --ipsource flag in the following code:
`````
if (defined($::opt_S)) {
        if (defined($bmc_user) && defined($bmc_pass) && defined($::opt_I)) {
            my $res = get_bmc_ip_source($::opt_I, $bmc_user, $bmc_pass);
            return $res;
        }
``````

The fixes are
1) initialized the bmc_user if $::opt_U is set
2) initialized the bmc_user if $::opt_P is set
3) command will return more meanfull error message

### CODE REVIEW

**Comment by cxhong** [comment]
@zet809 can u review this?

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm [hunk 1]
```diff
@@ -204,6 +204,16 @@ sub bmcdiscovery_processargs {
         return 1;
     }
 
+    ############################################
+    # Option -U and -P for bmc user and password 
+    ############################################
+    if ($::opt_U) {
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: I think you shall follow line 243-254 to deal with the $bmc_user and $bmc_password.

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm [hunk 2]
```diff
@@ -303,6 +313,8 @@ sub bmcdiscovery_processargs {
                 $msg = "The ipsource option requires a BMC IP.  Specify the IP using the -i|--bmcip option.";
             } elsif (!defined($::opt_P)) {
                 $msg = "The ipsource option requires a password.  Specify the password with the -p|--bmcpasswd option.";
+            } else {
+                $msg = "Failed to process ipsource command for bmc ip=$::opt_I user=$bmc_user password=$bmc_pass";
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: The $::opt_U can be not specified for Tuleta server, they won't have bmcusername enabled.

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm (hunk 1 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: I will remove out line 243-254 code in here.  so will not have duplicate code in one subroutine

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm (hunk 2 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: will remove checking for (defined($bmc_user)) on if statement

**Comment by zet809** [comment]
@bybai Hi, bai yuan, will you pls help to review this pull request?

**Comment by bybai** [comment]
Hi @cxhong , could you please remove the useless blank lines, such as line 351,358 360 etc?

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm [hunk 3]
```diff
@@ -204,6 +204,25 @@ sub bmcdiscovery_processargs {
...
+    # Option -U and -P for bmc user and password 
+    #
+    # Get the default bmc account from passwd table, 
+    # this is only done for the discovery process
+    ############################################
+    ($bmc_user, $bmc_pass) = bmcaccount_from_passwd();
+    # overwrite the default user and password if one is provided
+    if ($::opt_U) {
+        $bmc_user = $::opt_U;
+    } elsif ($::opt_P) {
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: Why is this line changed to check for opt_P?

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm (hunk 2 above)
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: Should we be checking that the result of get_bmc_ip_source comes back OK up above @ line 299?

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm (hunk 3 above)
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: Alright, I see.. , I'm ok with this.

File: xCAT-server/lib/xcat/plugins/bmcdiscover.pm (hunk 2 above)
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/bmcdiscover.pm]: Alright, talked to @cxhong, This is okay.

**Comment by whowutwut** [comment]
I've reviewed the change and will merge.
